src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/ProjectSnapshotManagerBase.cs

```diff
@@ -11,6 +12,15 @@ internal abstract class ProjectSnapshotManagerBase : ProjectSnapshotManager
 
         public abstract void DocumentAdded(HostProject hostProject, HostDocument hostDocument);
 
+        // Yeah this is kinda ugly.
+        public abstract void DocumentOpened(string projectFilePath, string documentFilePath, SourceTextContainer textContainer);
+
+        public abstract void DocumentClosed(string projectFilePath, string documentFilePath, TextLoader textLoader);
+
+        public abstract void DocumentChanged(string projectFilePath, string documentFilePath, TextLoader textLoader);
+
+        public abstract void DocumentChanged(string projectFilePath, string documentFilePath, SourceText sourceText);
```

Comment:
I don't love the design of these APIs. I feel like we badly want something like `ProjectId` `DocumentId`. It might be time to just start using them.